rfcs/fast-n-furious.md

```diff
@@ -0,0 +1,354 @@
+# Improvements to a compiler to make it fast ⚡️
+
+_Authors: [@layershifter](https://github.com/layershifter)_
+
+Griffel features a compiler (available as a Webpack loader, Vite plugin, etc.) that executes [Ahead-of-Time (AOT) compilation](https://griffel.js.org/react/ahead-of-time-compilation/introduction) to enhance runtime performance and shift workload to the build phase. The compiler also enables [CSS extraction](https://griffel.js.org/react/css-extraction/introduction), offering significant performance benefits for applications.
+
+However, we got multiple reports indicate that the compiler's speed is suboptimal. This project seeks to enhance the compiler's performance.
+
+## What the compiler does?
+
+The compiler processes `makeStyles()` & `makeResetStyles()` calls to pre-create CSS rules. For instance:
+
+```jsx
+// Input code
+
+const useStyles = makeStyles({
+  root: {
+    paddingLeft: '1px',
+    display: 'flex',
+  },
+});
+```
+
+```jsx
+// Output code
+
+const useStyles = __styles(
+  {
+    root: {
+      mc9l5x: 'f22iagw',
+      uwmqm3: ['f10xn8zz', 'f136y8j8'],
+    },
+  },
+  {
+    d: ['.f22iagw{display:flex;}', '.f10xn8zz{padding-left:1px;}', '.f136y8j8{padding-right:1px;}'],
+  },
+);
+```
+
+See [Griffel.js: Technical details about AOT](https://griffel.js.org/react/ahead-of-time-compilation/technical-details) to get the complete overview.
+
+## How it works?
+
+The complexity in the compiler arises when a `makeStyles()` call includes an imported value from another module. For example:
+
+```jsx
+/** @filename `input.ts` */
+
+import { makeStyles } from '@griffel/react';
+import { spacing } from './theme';
+
+const useStyles = makeStyles({
+  root: { paddingLeft: spacing(1) },
+});
+```
+
+In this scenario, the context of `input.ts` is insufficient to pre-create CSS rules. The compiler must resolve the `spacing()` function and substitute its call with the actual value.
+
+This introduces complexity. Internally, the compiler functions as a micro-bundler that resolves imports, evaluates expressions, and replaces them with actual values. The simplified lifecycle is below.
+
+```mermaid
+graph TD
```

Comment:
code snippets in diagrams seem to be broken for me 
![image](https://github.com/user-attachments/assets/ded9c9c6-8ce8-4cd9-8ed3-1f51c9bd45f8)